### ISSUE TITLE
Download/upload progress?

### ISSUE THREAD

**Issue by AxelTerizaki** [patrickjuchli/basic-ftp#6]
I suppose this is something that could be done externally by analyzing the stream we write/read to/from, but it might be nice to have a way to track progress of oa download or upload in basic-ftp.

BTW, very nice module, it's the easiest I found for downloading a bunch of files sequentially with the same connection, and it works fine with promises :)

**Comment by patrickjuchli** [comment]
Thanks for the kind words! 

Some progress information seems reasonable but I have to think about this for a bit. The library should do as little as possible and whatever it is doing should work well for directory-level methods as well.

**Comment by patrickjuchli** [comment]
Okay, progress reporting will probably look something like the following example and be part of the next release.

```
// Log progress for any transfer from now on.
client.trackProgress(info => {
    console.log("File", info.name);
    console.log("Type", info.transfer);
    console.log("Single/current transfer", info.bytes);
    console.log("All transfers", info.bytesOverall);
});

// Transfer some data
await client.upload(someStream, "test.txt");
await client.upload(someOtherStream, "test2.txt");

// Reset overall counter 
client.trackProgress(info => console.log(info.bytesOverall));
client.downloadDir("local/path");

// Stop logging
client.trackProgress();
```

**Comment by patrickjuchli** [comment]
The library will not take care of showing a percentage. To do this you'll have to determine the size of the file (or the directory) yourself and then provide a corresponding handler function for `trackProgress`.

**Comment by AxelTerizaki** [comment]
So if I understand correctly, the function handler for `trackProgress` will be called each time there's a change in progress of a file? Each tiem there's progress, the fonction is called? It seems easy this way, just have to update a progress bar.

Thanks for your hard work :)

**Comment by patrickjuchli** [comment]
Yes. You can call `trackProgress` once at the beginning and then display progress for all following transfers. The handler function will also receive the name of the current file (the remote file name to be exact) and whether it's an upload or download.

You can also call `trackProgress` again later to reset bytesOverall. BytesOverall counts all bytes transferred by all transfers so far. This is useful if you want to provide total downloaded bytes when downloading a directory with many files for example.

You stop logging progress by just passing an undefined handler function to `trackProgress`.

One more detail: As soon as a transfer started, the progress handler function is called at a regular interval (probably 500ms) and not just when data chunks are transferred. This makes it easier for you to e.g. derive a transfer speed because you get updates even if transfer slows down to 0 during an upload.

**Comment by patrickjuchli** [comment]
But be aware that in order to be able to show a progress bar, you need the fraction/percentage of the transfer. This library only provides the number of bytes transferred. It is up to you to get the total size of whatever you transfer and then derive a percentage.

**Comment by patrickjuchli** [comment]
Released in 2.9.0
